


### Description of files
All these files are used by the jupyter/ipython notebook [plot_ml_trees_with_metadata.ipynb](plot_ml_trees_with_metadata.ipynb) for generating the figure.
Output files (generated by the notebook) have prefix `output`. 

Files [01.mlst_genomes.tsv](01.mlst_genomes.tsv) and [01.mlst_mags.tsv](01.mlst_mags.tsv) contain the MLST results for the genomes and metagenome-assembled genomes (MAGs) in the dataset, respectively. 

File [02.amrfinder.wide.tsv](02.amrfinder.wide.tsv) contains the AMR gene detection results for the genomes and MAGs in the dataset. 
It was generated using the AMRFinderPlus tool, and a short R script to merge results from multiple files: `amrwide <- amr %>% mutate(value="YES") %>% spread (key=Gene.symbol, value=value, fill="NO")`

Files [03.GTDBTk.isolates.names_species.tsv](03.GTDBTk.isolates.names_species.tsv), [03.GTDBTk.mags.names_species.tsv](03.GTDBTk.mags.names_species.tsv), and 
[03.GTDBTk.mags_new.names_species.tsv](03.GTDBTk.mags_new.names_species.tsv) contain the taxonomic classification of the genomes and MAGs in the dataset.
There were two batches of MAGs, which explain the two files. 

File [04.name_mapping.csv](04.name_mapping.csv) contains the mapping between file names, display names, etc for the MAGs 

File [05.trimal99.aln.treefile](05.trimal99.aln.treefile) contains the phylogenetic tree generated from the panaroo+bakta alignment after trimming with trimal at 1% indel-gap threshold.
In detail, for a generic case:
```
nohup panaroo -i 00.60_genomes_bakta/PID_* ../12.all_19_metagenomes/*gff3 -o panaroo_MAGS_and_60genomes-core --remove-invalid-genes -t 38 --core_threshold 0.2 -a core  --clean-mode strict &
trimal -in ../panaroo_MAGS_and_60genomes-core/core_gene_alignment.aln -out filtered_trimal_bt99.aln  -gt 0.99
nohup sh -c 'iqtree -s filtered_trimal_bt99.snps.aln -m HKY+G -t core_gene_alignment.snps.aln.treefile -nt 12' > nohup.out  &
```

File [06.gene_presence_absence.Rtab](06.gene_presence_absence.Rtab) contains the gene presence/absence matrix generated by panaroo for the genomes and MAGs in the dataset and is used only for checking some statistics (see [plot_ml_trees_with_metadata.ipynb](plot_ml_trees_with_metadata.ipynb)).
File [06.aligned_gene_sequences-length.tsv](06.aligned_gene_sequences-length.tsv) contains the alignment length and number of sequences for each gene in the panaroo output, and was generated with a bash command including `goalign stats length`. It's also used only for showing the pangenome statistics...


